### PR TITLE
[FLINK-22881] Tasks are blocked while broadcasting stream status

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

We revert the toggling between IDLE and ACTIVE status if a record is emitted in an IDLE state. If records are emitted in an IDLE state the toggling incur a lot of performance penalty.

Moreover I revert the fact that we do not mark output as idle if there are no `PartialWatermarks`.

Lastly, I optimize the `WatermarkAssignerOperator` slightly so that it emits `StreamStatus` only if the announcedStatus changed.

The last commit was reviewed in https://github.com/apache/flink/pull/16095

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by lirui-apache** [comment]
Hey @dawidwys , thanks for taking care of this. The documentation about `ExecutionConfigOptions::TABLE_EXEC_SOURCE_IDLE_TIMEOUT` says by default the idleness detect is not enabled. So I wonder why we need to set it for the hive test?

**Comment by dawidwys** [comment]
@lirui-apache The idea was that if we enable `TABLE_EXEC_SOURCE_IDLE_TIMEOUT` the `WatermarkAssignerOperator` takes control over its `StreamStatus` and it does not respect the idleness of the sources.

However, I spotted another difference from the 1.13 in how idleness is handled in sources which I reverted in this PR.

**Comment by lirui-apache** [comment]
Thanks @dawidwys for the clarifications. I'm not really familiar with the watermark and idleness detection code. So I'll leave this for other folks to have a look.

**Comment by dawidwys** [comment]
@pnowojski @AHeise One question I have is if I shall put the logic that the multiplexer can go idle after splits are removed (watermark has progressed from its initial value) or should it be active in that case?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java [hunk 1]
```diff
@@ -652,11 +652,11 @@ private void processStreamStatus(StreamStatus streamStatus, int index) throws Ex
         }
     }
 
-    public final void processStreamStatus1(StreamStatus streamStatus) throws Exception {
+    public void processStreamStatus1(StreamStatus streamStatus) throws Exception {
         processStreamStatus(streamStatus, 0);
     }
 
-    public final void processStreamStatus2(StreamStatus streamStatus) throws Exception {
+    public void processStreamStatus2(StreamStatus streamStatus) throws Exception {
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java:659]: Are these changes still needed?

File: flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermarkStatus.java [hunk 2]
```diff
@@ -45,7 +45,7 @@ public long getCombinedWatermark() {
     }
 
     public boolean isIdle() {
-        return idle;
+        return (!partialWatermarks.isEmpty() || combinedWatermark > Long.MIN_VALUE) && idle;
```
Comment (AHeise) [inline flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermarkStatus.java:48]: Couldn't we move the additional checks to `updateCombinedWatermark` to your newly added block? See suggestion below.

File: flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermarkStatus.java [hunk 3]
```diff
@@ -66,25 +66,25 @@ public void add(PartialWatermark element) {
     public boolean updateCombinedWatermark() {
         long minimumOverAllOutputs = Long.MAX_VALUE;
 
-        boolean hasOutputs = false;
+        // if we don't have any outputs minimumOverAllOutputs is not valid, it's still
+        // at its initial Long.MAX_VALUE state and we must not emit that
+        if (partialWatermarks.isEmpty()) {
+            this.idle = true;
```
Comment (AHeise) [inline flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermarkStatus.java:72]: `this.idle = combinedWatermark > Long.MIN_VALUE`

**Review by AHeise — APPROVED** [review]
LGTM with 2 suggestions.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java (hunk 1 above)
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java:659]: You mean the entire commit? I think it makes sense, otherwise the operator will generate `StreamStatus` for every record.

File: flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermarkStatus.java (hunk 3 above)
Comment (dawidwys) [reply flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermarkStatus.java:72]: good idea!

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java (hunk 1 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java:659]: No I meant the removal of `final` in `processStreamStatus1/2`. Doesn't seem to be necessary or the fix is incomplete.

Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java:659]: I override these two methods in `WatermarkAssignerOperator` which extends from `AbstractStreamOperator`. Therefore I need to remove the `final` here.

Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java:659]: Scratch that. I need to remove `final` only from `processStreamStatus`.

**Comment by dawidwys** [comment]
I'll wait for green CI and merge it. Thx @AHeise for the review.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java (hunk 1 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: That's what I meant :p
